{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/sanjaynagi/AnoFold/blob/main/notebooks/AnoFold.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "x5wID6CTbzRB",
    "outputId": "7fefba6d-9d1b-4acf-ee5b-3c3e10aa7339"
   },
   "outputs": [],
   "source": [
    "# %pip install py3Dmol\n",
    "\n",
    "# !wget https://github.com/gnina/gnina/releases/download/v1.1/gnina\n",
    "# !chmod +x gnina\n",
    "# !./gnina --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "XVXkg5Xi8bdU"
   },
   "outputs": [],
   "source": [
    "# # genes = ['AGAP006227', 'AGAP006228', 'AGAP006723', 'AGAP006724', 'AGAP006725', 'AGAP006726', 'AGAP006727']\n",
    "# df_genes = pd.read_csv(\"https://raw.githubusercontent.com/sanjaynagi/AnoExpress/main/resources/AgamP4.annots.tsv\", \n",
    "#                        sep=\"\\t\")\n",
    "\n",
    "# search = 'S-transferase'\n",
    "# genes = df_genes.fillna('bl').query('GeneDescription.str.contains(@search)')['GeneID'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "XVXkg5Xi8bdU"
   },
   "outputs": [],
   "source": [
    "# ligands = ['ddt3d', 'deltamethrin3d']           #, 'pirimiphos-methyl-oxon3d', 'malathion3d', 'cis-permethrin3d','transfluthrin3d']\n",
    "# genes = ['AGAP006227']\n",
    "# # for gene in genes:\n",
    "# #     for ligand in ligands:\n",
    "# #           res = invin(gene, ligand=ligand, override_desc='glutathione transferase')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "FmfNINymb3_o"
   },
   "source": [
    "## iMOVE - inverse molecular docking with alphafold and gnina"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import imove\n",
    "\n",
    "mutagen_dicts = [None, {114:'T'}, {119:'V'}, {120:'L'}]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting docking process for gene_id: AGAP009194, ligand: ddt3d\n",
      "Checking for receptor and ligand files...\n",
      "UniProt accession: K7WDM7\n",
      "Running molecular docking for AGAP009194 | GSTe2 (glutathione transferase)\n",
      "Generated motifs - active site motif: RFIFERI, catalytic molecule: O, catalytic codon in motif: 4\n",
      "Motif detected in PDB at codon 112:118 = RFIFERI, Flanking sequence: LFARM[RFIFERI]LLFGK\n",
      "Active site coordinates: x=17.031, y=2.467, z=-4.572\n",
      "Skipping GNINA docking: log file already exists or ligand file is missing\n",
      "Docking process completed\n",
      "Starting docking process for gene_id: AGAP009194, ligand: ddt3d\n",
      "Checking for receptor and ligand files...\n",
      "Using provided override motif - active site motif: RFTFERI, catalytic molecule: O, catalytic codon in motif: 4\n",
      "Motif detected in PDB at codon 112:118 = RFTFERI, Flanking sequence: LFARM[RFTFERI]LLFGK\n",
      "Active site coordinates: x=17.031, y=2.467, z=-4.572\n",
      "Preparing to run GNINA for docking...\n",
      "              _             \n",
      "             (_)            \n",
      "   __ _ _ __  _ _ __   __ _ \n",
      "  / _` | '_ \\| | '_ \\ / _` |\n",
      " | (_| | | | | | | | | (_| |\n",
      "  \\__, |_| |_|_|_| |_|\\__,_|\n",
      "   __/ |                    \n",
      "  |___/                     \n",
      "\n",
      "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n",
      "gnina is based on smina and AutoDock Vina.\n",
      "Please cite appropriately.\n",
      "\n",
      "WARNING: No GPU detected. CNN scoring will be slow.\n",
      "Recommend running with single model (--cnn crossdock_default2018)\n",
      "or without cnn scoring (--cnn_scoring=none).\n",
      "\n",
      "Commandline: ./gnina -r ../../receptors/AGAP009194_114T.pdbqt -l ../../ligands/ddt3d.pdbqt --center_x 17.031 --center_y 2.467 --center_z -4.572 --size_x 20 --size_y 20 --size_z 20 -o ../../docked/AGAP009194_114T_ddt3d.sdf --log ../../logs/AGAP009194_114T_ddt3d.log --seed 0\n",
      "Using random seed: 0\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      " | pose 0 | initial pose not within box\n",
      "\n",
      "mode |  affinity  |  intramol  |    CNN     |   CNN\n",
      "     | (kcal/mol) | (kcal/mol) | pose score | affinity\n",
      "-----+------------+------------+------------+----------\n",
      "    1       -8.31       -0.37       0.8366      7.209\n",
      "    2       -8.06       -0.22       0.8141      7.083\n",
      "    3       -8.03       -0.45       0.8140      7.075\n",
      "    4       -8.00       -0.12       0.7949      7.164\n",
      "    5       -7.83       -0.27       0.7910      7.126\n",
      "    6       -3.03       -0.43       0.5720      4.734\n",
      "    7       -4.59       -0.09       0.3904      4.472\n",
      "    8       -4.43       -0.41       0.3809      4.566\n",
      "    9       -3.41       -0.68       0.3736      4.508\n",
      "Docking process completed\n",
      "Starting docking process for gene_id: AGAP009194, ligand: ddt3d\n",
      "Checking for receptor and ligand files...\n",
      "Receptor file not found. Downloading and preparing AlphaFold PDB for AGAP009194\n",
      "┌──────────────────────────────────────────────────────────────────────────────┐\n",
      "│                                 PyRosetta-4                                  │\n",
      "│              Created in JHU by Sergey Lyskov and PyRosetta Team              │\n",
      "│              (C) Copyright Rosetta Commons Member Institutions               │\n",
      "│                                                                              │\n",
      "│ NOTE: USE OF PyRosetta FOR COMMERCIAL PURPOSES REQUIRE PURCHASE OF A LICENSE │\n",
      "│         See LICENSE.PyRosetta.md or email license@uw.edu for details         │\n",
      "└──────────────────────────────────────────────────────────────────────────────┘\n",
      "PyRosetta-4 2024 [Rosetta PyRosetta4.Release.python311.ubuntu 2024.19+release.a34b73c40fe9c61558d566d6a63f803cfb15a4fc 2024-05-02T16:22:03] retrieved from: http://www.pyrosetta.org\n",
      "core.init: Checking for fconfig files in pwd and ./rosetta/flags\n",
      "core.init: Rosetta version: PyRosetta4.Release.python311.ubuntu r381 2024.19+release.a34b73c40f a34b73c40fe9c61558d566d6a63f803cfb15a4fc http://www.pyrosetta.org 2024-05-02T16:22:03\n",
      "core.init: Rosetta extras: []\n",
      "core.init: command: PyRosetta -write_pdb_title_section_records true -write_pdb_link_records true -use_pdb_format_HETNAM_records true -write_pdb_parametric_info true -write_glycan_pdb_codes false -write_seqres_records true -output_pose_energies_table false -database /home/sanj/apps/mambaforge/lib/python3.11/site-packages/pyrosetta/database\n",
      "basic.random.init_random_generator: 'RNG device' seed mode, using '/dev/urandom', seed=-1273262164 seed_offset=0 real_seed=-1273262164\n",
      "basic.random.init_random_generator: RandomGenerator:init: Normal mode, seed=-1273262164 RG_type=mt19937\n",
      "core.import_pose.import_pose: File '../../receptors/AGAP009194_raw.pdb' automatically determined to be of type PDB\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "core.pack.task: Packer task: initialize from command line()\n",
      "core.pack.pack_rotamers: built 1 rotamers at 1 positions.\n",
      "core.pack.interaction_graph.interaction_graph_factory: Instantiating PDInteractionGraph\n",
      "Mutated residue 119 to V\n",
      "Mutagenesis complete. New structure saved as '../../receptors/AGAP009194_raw_119V.pdb'\n",
      "Downloaded and protonated AlphaFold PDB for AGAP009194 (UniProt: K7WDM7) to ../../receptors/AGAP009194_119V.pdbqt\n",
      "UniProt accession: K7WDM7\n",
      "Running molecular docking for AGAP009194 | GSTe2 (glutathione transferase)\n",
      "Generated motifs - active site motif: RFIFERI, catalytic molecule: O, catalytic codon in motif: 4\n",
      "Motif detected in PDB at codon 112:118 = RFIFERI, Flanking sequence: LFARM[RFIFERI]VLFGK\n",
      "Active site coordinates: x=17.031, y=2.467, z=-4.572\n",
      "Preparing to run GNINA for docking...\n",
      "              _             \n",
      "             (_)            \n",
      "   __ _ _ __  _ _ __   __ _ \n",
      "  / _` | '_ \\| | '_ \\ / _` |\n",
      " | (_| | | | | | | | | (_| |\n",
      "  \\__, |_| |_|_|_| |_|\\__,_|\n",
      "   __/ |                    \n",
      "  |___/                     \n",
      "\n",
      "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n",
      "gnina is based on smina and AutoDock Vina.\n",
      "Please cite appropriately.\n",
      "\n",
      "WARNING: No GPU detected. CNN scoring will be slow.\n",
      "Recommend running with single model (--cnn crossdock_default2018)\n",
      "or without cnn scoring (--cnn_scoring=none).\n",
      "\n",
      "Commandline: ./gnina -r ../../receptors/AGAP009194_119V.pdbqt -l ../../ligands/ddt3d.pdbqt --center_x 17.031 --center_y 2.467 --center_z -4.572 --size_x 20 --size_y 20 --size_z 20 -o ../../docked/AGAP009194_119V_ddt3d.sdf --log ../../logs/AGAP009194_119V_ddt3d.log --seed 0\n",
      "Using random seed: 0\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      " | pose 0 | initial pose not within box\n",
      " | pose 0 | ligand outside box\n",
      "\n",
      "mode |  affinity  |  intramol  |    CNN     |   CNN\n",
      "     | (kcal/mol) | (kcal/mol) | pose score | affinity\n",
      "-----+------------+------------+------------+----------\n",
      "    1       -7.81       -0.05       0.8112      7.040\n",
      "    2       -8.03       -0.60       0.7661      6.964\n",
      "    3       -8.19       -0.49       0.7540      7.090\n",
      "    4       -8.18       -0.49       0.7492      7.082\n",
      "    5       -8.11       -0.36       0.6571      6.778\n",
      "    6       -8.07       -0.68       0.6442      6.898\n",
      "    7       -3.65       -0.19       0.4699      4.461\n",
      "    8       -3.97       -0.55       0.4182      4.785\n",
      "    9       -4.46       -0.26       0.3735      4.555\n",
      "Docking process completed\n",
      "Starting docking process for gene_id: AGAP009194, ligand: ddt3d\n",
      "Checking for receptor and ligand files...\n",
      "Receptor file not found. Downloading and preparing AlphaFold PDB for AGAP009194\n",
      "┌──────────────────────────────────────────────────────────────────────────────┐\n",
      "│                                 PyRosetta-4                                  │\n",
      "│              Created in JHU by Sergey Lyskov and PyRosetta Team              │\n",
      "│              (C) Copyright Rosetta Commons Member Institutions               │\n",
      "│                                                                              │\n",
      "│ NOTE: USE OF PyRosetta FOR COMMERCIAL PURPOSES REQUIRE PURCHASE OF A LICENSE │\n",
      "│         See LICENSE.PyRosetta.md or email license@uw.edu for details         │\n",
      "└──────────────────────────────────────────────────────────────────────────────┘\n",
      "PyRosetta-4 2024 [Rosetta PyRosetta4.Release.python311.ubuntu 2024.19+release.a34b73c40fe9c61558d566d6a63f803cfb15a4fc 2024-05-02T16:22:03] retrieved from: http://www.pyrosetta.org\n",
      "core.init: Checking for fconfig files in pwd and ./rosetta/flags\n",
      "core.init: Rosetta version: PyRosetta4.Release.python311.ubuntu r381 2024.19+release.a34b73c40f a34b73c40fe9c61558d566d6a63f803cfb15a4fc http://www.pyrosetta.org 2024-05-02T16:22:03\n",
      "core.init: Rosetta extras: []\n",
      "core.init: command: PyRosetta -write_pdb_title_section_records true -write_pdb_link_records true -use_pdb_format_HETNAM_records true -write_pdb_parametric_info true -write_glycan_pdb_codes false -write_seqres_records true -output_pose_energies_table false -database /home/sanj/apps/mambaforge/lib/python3.11/site-packages/pyrosetta/database\n",
      "basic.random.init_random_generator: 'RNG device' seed mode, using '/dev/urandom', seed=-376989801 seed_offset=0 real_seed=-376989801\n",
      "basic.random.init_random_generator: RandomGenerator:init: Normal mode, seed=-376989801 RG_type=mt19937\n",
      "core.import_pose.import_pose: File '../../receptors/AGAP009194_raw.pdb' automatically determined to be of type PDB\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "core.pack.task: Packer task: initialize from command line()\n",
      "core.pack.pack_rotamers: built 2 rotamers at 1 positions.\n",
      "core.pack.interaction_graph.interaction_graph_factory: Instantiating DensePDInteractionGraph\n",
      "Mutated residue 120 to L\n",
      "Mutagenesis complete. New structure saved as '../../receptors/AGAP009194_raw_120L.pdb'\n",
      "Downloaded and protonated AlphaFold PDB for AGAP009194 (UniProt: K7WDM7) to ../../receptors/AGAP009194_120L.pdbqt\n",
      "UniProt accession: K7WDM7\n",
      "Running molecular docking for AGAP009194 | GSTe2 (glutathione transferase)\n",
      "Generated motifs - active site motif: RFIFERI, catalytic molecule: O, catalytic codon in motif: 4\n",
      "Motif detected in PDB at codon 112:118 = RFIFERI, Flanking sequence: LFARM[RFIFERI]LLFGK\n",
      "Active site coordinates: x=17.031, y=2.467, z=-4.572\n",
      "Preparing to run GNINA for docking...\n",
      "              _             \n",
      "             (_)            \n",
      "   __ _ _ __  _ _ __   __ _ \n",
      "  / _` | '_ \\| | '_ \\ / _` |\n",
      " | (_| | | | | | | | | (_| |\n",
      "  \\__, |_| |_|_|_| |_|\\__,_|\n",
      "   __/ |                    \n",
      "  |___/                     \n",
      "\n",
      "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n",
      "gnina is based on smina and AutoDock Vina.\n",
      "Please cite appropriately.\n",
      "\n",
      "WARNING: No GPU detected. CNN scoring will be slow.\n",
      "Recommend running with single model (--cnn crossdock_default2018)\n",
      "or without cnn scoring (--cnn_scoring=none).\n",
      "\n",
      "Commandline: ./gnina -r ../../receptors/AGAP009194_120L.pdbqt -l ../../ligands/ddt3d.pdbqt --center_x 17.031 --center_y 2.467 --center_z -4.572 --size_x 20 --size_y 20 --size_z 20 -o ../../docked/AGAP009194_120L_ddt3d.sdf --log ../../logs/AGAP009194_120L_ddt3d.log --seed 0\n",
      "Using random seed: 0\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      " | pose 0 | initial pose not within box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | initial pose not within box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      "\n",
      "mode |  affinity  |  intramol  |    CNN     |   CNN\n",
      "     | (kcal/mol) | (kcal/mol) | pose score | affinity\n",
      "-----+------------+------------+------------+----------\n",
      "    1       -8.29       -0.55       0.8342      7.210\n",
      "    2       -8.02       -0.19       0.8196      7.097\n",
      "    3       -8.00       -0.30       0.8135      7.085\n",
      "    4       -7.82       -0.18       0.7698      7.123\n",
      "    5       -3.67       -0.63       0.5704      4.538\n",
      "    6       -3.63       -0.42       0.5519      4.550\n",
      "    7       -3.97       -0.45       0.5140      4.420\n",
      "    8        1.38       -0.03       0.4765      4.709\n",
      "    9       -7.11       -0.40       0.4119      6.443\n",
      "Docking process completed\n",
      "Starting docking process for gene_id: AGAP009194, ligand: deltamethrin3d\n",
      "Checking for receptor and ligand files...\n",
      "UniProt accession: K7WDM7\n",
      "Running molecular docking for AGAP009194 | GSTe2 (glutathione transferase)\n",
      "Generated motifs - active site motif: RFIFERI, catalytic molecule: O, catalytic codon in motif: 4\n",
      "Motif detected in PDB at codon 112:118 = RFIFERI, Flanking sequence: LFARM[RFIFERI]LLFGK\n",
      "Active site coordinates: x=17.031, y=2.467, z=-4.572\n",
      "Preparing to run GNINA for docking...\n",
      "              _             \n",
      "             (_)            \n",
      "   __ _ _ __  _ _ __   __ _ \n",
      "  / _` | '_ \\| | '_ \\ / _` |\n",
      " | (_| | | | | | | | | (_| |\n",
      "  \\__, |_| |_|_|_| |_|\\__,_|\n",
      "   __/ |                    \n",
      "  |___/                     \n",
      "\n",
      "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n",
      "gnina is based on smina and AutoDock Vina.\n",
      "Please cite appropriately.\n",
      "\n",
      "WARNING: No GPU detected. CNN scoring will be slow.\n",
      "Recommend running with single model (--cnn crossdock_default2018)\n",
      "or without cnn scoring (--cnn_scoring=none).\n",
      "\n",
      "Commandline: ./gnina -r ../../receptors/AGAP009194.pdbqt -l ../../ligands/deltamethrin3d.pdbqt --center_x 17.031 --center_y 2.467 --center_z -4.572 --size_x 20 --size_y 20 --size_z 20 -o ../../docked/AGAP009194_deltamethrin3d.sdf --log ../../logs/AGAP009194_deltamethrin3d.log --seed 0\n",
      "Using random seed: 0\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      " | pose 0 | initial pose not within box\n",
      "\n",
      "mode |  affinity  |  intramol  |    CNN     |   CNN\n",
      "     | (kcal/mol) | (kcal/mol) | pose score | affinity\n",
      "-----+------------+------------+------------+----------\n",
      "    1       -7.85       -0.34       0.8093      7.106\n",
      "    2       -7.12        1.18       0.7540      6.875\n",
      "    3       -5.10       -1.03       0.6418      5.477\n",
      "    4       -6.13        0.55       0.5264      6.388\n",
      "    5       -6.24        1.23       0.4730      6.323\n",
      "    6       -4.54       -1.18       0.4611      5.209\n",
      "    7       -4.72       -0.38       0.4237      5.045\n",
      "    8       -5.07       -0.79       0.4227      5.190\n",
      "    9       -4.71        0.03       0.4166      5.318\n",
      "Docking process completed\n",
      "Starting docking process for gene_id: AGAP009194, ligand: deltamethrin3d\n",
      "Checking for receptor and ligand files...\n",
      "Using provided override motif - active site motif: RFTFERI, catalytic molecule: O, catalytic codon in motif: 4\n",
      "Motif detected in PDB at codon 112:118 = RFTFERI, Flanking sequence: LFARM[RFTFERI]LLFGK\n",
      "Active site coordinates: x=17.031, y=2.467, z=-4.572\n",
      "Preparing to run GNINA for docking...\n",
      "              _             \n",
      "             (_)            \n",
      "   __ _ _ __  _ _ __   __ _ \n",
      "  / _` | '_ \\| | '_ \\ / _` |\n",
      " | (_| | | | | | | | | (_| |\n",
      "  \\__, |_| |_|_|_| |_|\\__,_|\n",
      "   __/ |                    \n",
      "  |___/                     \n",
      "\n",
      "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n",
      "gnina is based on smina and AutoDock Vina.\n",
      "Please cite appropriately.\n",
      "\n",
      "WARNING: No GPU detected. CNN scoring will be slow.\n",
      "Recommend running with single model (--cnn crossdock_default2018)\n",
      "or without cnn scoring (--cnn_scoring=none).\n",
      "\n",
      "Commandline: ./gnina -r ../../receptors/AGAP009194_114T.pdbqt -l ../../ligands/deltamethrin3d.pdbqt --center_x 17.031 --center_y 2.467 --center_z -4.572 --size_x 20 --size_y 20 --size_z 20 -o ../../docked/AGAP009194_114T_deltamethrin3d.sdf --log ../../logs/AGAP009194_114T_deltamethrin3d.log --seed 0\n",
      "Using random seed: 0\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      " | pose 0 | initial pose not within box\n",
      "\n",
      "mode |  affinity  |  intramol  |    CNN     |   CNN\n",
      "     | (kcal/mol) | (kcal/mol) | pose score | affinity\n",
      "-----+------------+------------+------------+----------\n",
      "    1       -7.84       -0.47       0.8232      7.164\n",
      "    2       -7.28       -0.40       0.7466      6.873\n",
      "    3       -7.23        1.19       0.6455      6.741\n",
      "    4       -7.89       -0.27       0.6296      6.957\n",
      "    5       -6.07        1.01       0.5385      6.740\n",
      "    6       -5.52       -0.34       0.5288      5.237\n",
      "    7       -5.64        0.33       0.4994      6.389\n",
      "    8       -6.35       -0.97       0.4746      6.851\n",
      "    9       -5.38        0.58       0.4583      6.413\n",
      "Docking process completed\n",
      "Starting docking process for gene_id: AGAP009194, ligand: deltamethrin3d\n",
      "Checking for receptor and ligand files...\n",
      "UniProt accession: K7WDM7\n",
      "Running molecular docking for AGAP009194 | GSTe2 (glutathione transferase)\n",
      "Generated motifs - active site motif: RFIFERI, catalytic molecule: O, catalytic codon in motif: 4\n",
      "Motif detected in PDB at codon 112:118 = RFIFERI, Flanking sequence: LFARM[RFIFERI]VLFGK\n",
      "Active site coordinates: x=17.031, y=2.467, z=-4.572\n",
      "Preparing to run GNINA for docking...\n",
      "              _             \n",
      "             (_)            \n",
      "   __ _ _ __  _ _ __   __ _ \n",
      "  / _` | '_ \\| | '_ \\ / _` |\n",
      " | (_| | | | | | | | | (_| |\n",
      "  \\__, |_| |_|_|_| |_|\\__,_|\n",
      "   __/ |                    \n",
      "  |___/                     \n",
      "\n",
      "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n",
      "gnina is based on smina and AutoDock Vina.\n",
      "Please cite appropriately.\n",
      "\n",
      "WARNING: No GPU detected. CNN scoring will be slow.\n",
      "Recommend running with single model (--cnn crossdock_default2018)\n",
      "or without cnn scoring (--cnn_scoring=none).\n",
      "\n",
      "Commandline: ./gnina -r ../../receptors/AGAP009194_119V.pdbqt -l ../../ligands/deltamethrin3d.pdbqt --center_x 17.031 --center_y 2.467 --center_z -4.572 --size_x 20 --size_y 20 --size_z 20 -o ../../docked/AGAP009194_119V_deltamethrin3d.sdf --log ../../logs/AGAP009194_119V_deltamethrin3d.log --seed 0\n",
      "Using random seed: 0\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      " | pose 0 | initial pose not within box\n",
      "\n",
      "mode |  affinity  |  intramol  |    CNN     |   CNN\n",
      "     | (kcal/mol) | (kcal/mol) | pose score | affinity\n",
      "-----+------------+------------+------------+----------\n",
      "    1       -7.14        2.17       0.7879      6.910\n",
      "    2       -7.59       -0.53       0.7782      7.065\n",
      "    3       -7.89        1.46       0.7534      6.967\n",
      "    4       -7.88       -0.55       0.5941      6.874\n",
      "    5       -7.73        0.27       0.5786      6.619\n",
      "    6       -7.09       -0.55       0.4981      6.455\n",
      "    7       -7.62        0.05       0.4978      6.487\n",
      "    8       -7.15       -0.28       0.4839      6.407\n",
      "    9       -5.63       -0.08       0.4610      6.188\n",
      "Docking process completed\n",
      "Starting docking process for gene_id: AGAP009194, ligand: deltamethrin3d\n",
      "Checking for receptor and ligand files...\n",
      "UniProt accession: K7WDM7\n",
      "Running molecular docking for AGAP009194 | GSTe2 (glutathione transferase)\n",
      "Generated motifs - active site motif: RFIFERI, catalytic molecule: O, catalytic codon in motif: 4\n",
      "Motif detected in PDB at codon 112:118 = RFIFERI, Flanking sequence: LFARM[RFIFERI]LLFGK\n",
      "Active site coordinates: x=17.031, y=2.467, z=-4.572\n",
      "Preparing to run GNINA for docking...\n",
      "              _             \n",
      "             (_)            \n",
      "   __ _ _ __  _ _ __   __ _ \n",
      "  / _` | '_ \\| | '_ \\ / _` |\n",
      " | (_| | | | | | | | | (_| |\n",
      "  \\__, |_| |_|_|_| |_|\\__,_|\n",
      "   __/ |                    \n",
      "  |___/                     \n",
      "\n",
      "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n",
      "gnina is based on smina and AutoDock Vina.\n",
      "Please cite appropriately.\n",
      "\n",
      "WARNING: No GPU detected. CNN scoring will be slow.\n",
      "Recommend running with single model (--cnn crossdock_default2018)\n",
      "or without cnn scoring (--cnn_scoring=none).\n",
      "\n",
      "Commandline: ./gnina -r ../../receptors/AGAP009194_120L.pdbqt -l ../../ligands/deltamethrin3d.pdbqt --center_x 17.031 --center_y 2.467 --center_z -4.572 --size_x 20 --size_y 20 --size_z 20 -o ../../docked/AGAP009194_120L_deltamethrin3d.sdf --log ../../logs/AGAP009194_120L_deltamethrin3d.log --seed 0\n",
      "Using random seed: 0\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      " | pose 0 | initial pose not within box\n",
      "\n",
      "mode |  affinity  |  intramol  |    CNN     |   CNN\n",
      "     | (kcal/mol) | (kcal/mol) | pose score | affinity\n",
      "-----+------------+------------+------------+----------\n",
      "    1       -7.98       -0.42       0.8238      7.184\n",
      "    2       -7.51       -0.11       0.8220      6.997\n",
      "    3       -6.92        0.04       0.6439      6.678\n",
      "    4       -6.14        1.41       0.6091      6.631\n",
      "    5       -6.21       -0.36       0.5164      6.615\n",
      "    6       -6.49       -0.84       0.4713      6.830\n",
      "    7       -5.64        0.15       0.4378      5.958\n",
      "    8       -8.02       -0.75       0.4222      6.824\n",
      "    9       -6.62        2.30       0.4153      6.847\n",
      "Docking process completed\n"
     ]
    }
   ],
   "source": [
    "import imove\n",
    "\n",
    "for ligand in ['ddt3d', 'deltamethrin3d']:\n",
    "    for mutagen_dict in mutagen_dicts:\n",
    "        \n",
    "        # if I114T, the motif needs to be adjusted to include the new mutation\n",
    "        override_motif = None\n",
    "        if mutagen_dict:\n",
    "            if list(mutagen_dict.keys())[0] == 114:\n",
    "                override_motif = ('RFTFERI', \"O\", 4)\n",
    "\n",
    "        imove.dock(\"AGAP009194\", ligand, wkdir=\"../../\", verbose=True, mutagenesis_dict=mutagen_dict, override_motif=override_motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutagen_dicts[1]\n",
    "\n",
    "override_motif = ('RFTFERI', \"O\", 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# docked = imove.dock(\"AGAP000818\", \"ddt3d\", wkdir=\"../../\", verbose=True, mutagenesis_dict=mutagen_dicts[3], override_motif=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mode</th>\n",
       "      <th>Affinity</th>\n",
       "      <th>Intramol</th>\n",
       "      <th>CNN_score</th>\n",
       "      <th>CNN_affinity</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-8.29</td>\n",
       "      <td>-0.55</td>\n",
       "      <td>0.8342</td>\n",
       "      <td>7.210</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>-8.02</td>\n",
       "      <td>-0.19</td>\n",
       "      <td>0.8196</td>\n",
       "      <td>7.097</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>-8.00</td>\n",
       "      <td>-0.30</td>\n",
       "      <td>0.8135</td>\n",
       "      <td>7.085</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>-7.82</td>\n",
       "      <td>-0.18</td>\n",
       "      <td>0.7698</td>\n",
       "      <td>7.123</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>-3.67</td>\n",
       "      <td>-0.63</td>\n",
       "      <td>0.5704</td>\n",
       "      <td>4.538</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>-3.63</td>\n",
       "      <td>-0.42</td>\n",
       "      <td>0.5519</td>\n",
       "      <td>4.550</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>-3.97</td>\n",
       "      <td>-0.45</td>\n",
       "      <td>0.5140</td>\n",
       "      <td>4.420</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>1.38</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>0.4765</td>\n",
       "      <td>4.709</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>-7.11</td>\n",
       "      <td>-0.40</td>\n",
       "      <td>0.4119</td>\n",
       "      <td>6.443</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mode  Affinity  Intramol  CNN_score  CNN_affinity Distance\n",
       "0     1     -8.29     -0.55     0.8342         7.210     None\n",
       "1     2     -8.02     -0.19     0.8196         7.097     None\n",
       "2     3     -8.00     -0.30     0.8135         7.085     None\n",
       "3     4     -7.82     -0.18     0.7698         7.123     None\n",
       "4     5     -3.67     -0.63     0.5704         4.538     None\n",
       "5     6     -3.63     -0.42     0.5519         4.550     None\n",
       "6     7     -3.97     -0.45     0.5140         4.420     None\n",
       "7     8      1.38     -0.03     0.4765         4.709     None\n",
       "8     9     -7.11     -0.40     0.4119         6.443     None"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "docked.df"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "include_colab_link": true,
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
