{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/sanjaynagi/AnoFold/blob/main/notebooks/AnoFold.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "x5wID6CTbzRB",
    "outputId": "7fefba6d-9d1b-4acf-ee5b-3c3e10aa7339"
   },
   "outputs": [],
   "source": [
    "# %pip install py3Dmol\n",
    "\n",
    "# !wget https://github.com/gnina/gnina/releases/download/v1.1/gnina\n",
    "# !chmod +x gnina\n",
    "# !./gnina --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "XVXkg5Xi8bdU"
   },
   "outputs": [],
   "source": [
    "# # genes = ['AGAP006227', 'AGAP006228', 'AGAP006723', 'AGAP006724', 'AGAP006725', 'AGAP006726', 'AGAP006727']\n",
    "# df_genes = pd.read_csv(\"https://raw.githubusercontent.com/sanjaynagi/AnoExpress/main/resources/AgamP4.annots.tsv\", \n",
    "#                        sep=\"\\t\")\n",
    "\n",
    "# search = 'S-transferase'\n",
    "# genes = df_genes.fillna('bl').query('GeneDescription.str.contains(@search)')['GeneID'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "id": "XVXkg5Xi8bdU"
   },
   "outputs": [],
   "source": [
    "# ligands = ['ddt3d', 'deltamethrin3d']           #, 'pirimiphos-methyl-oxon3d', 'malathion3d', 'cis-permethrin3d','transfluthrin3d']\n",
    "# genes = ['AGAP006227']\n",
    "# # for gene in genes:\n",
    "# #     for ligand in ligands:\n",
    "# #           res = invin(gene, ligand=ligand, override_desc='glutathione transferase')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "FmfNINymb3_o"
   },
   "source": [
    "## iMOVE - inverse molecular docking with alphafold and gnina"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import imove\n",
    "\n",
    "mutagen_dicts = [None, {114:'T'}, {119:'V'}, {120:'L'}]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for ligand in ['ddt3d', 'deltamethrin3d']:\n",
    "#     for mutagen_dict in mutagen_dicts:\n",
    "        \n",
    "#         # if I114T, the motif needs to be adjusted to include the new mutation\n",
    "#         override_motif = None\n",
    "#         if mutagen_dict:\n",
    "#             if list(mutagen_dict.keys())[0] == 114:\n",
    "#                 override_motif = ('RFTFERI', \"O\", 4)\n",
    "\n",
    "#         imove.dock(\"AGAP009194\", ligand, wkdir=\"../../\", verbose=True, mutagenesis_dict=mutagen_dict, override_motif=override_motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;34mdocked\u001b[0m/         \u001b[01;34mfigures\u001b[0m/  \u001b[01;34mligands\u001b[0m/  pyproject.toml  \u001b[01;34mreceptors\u001b[0m/\n",
      "example_df.csv  \u001b[01;34mimove\u001b[0m/    \u001b[01;34mlogs\u001b[0m/     README.md       \u001b[01;34mresources\u001b[0m/\n"
     ]
    }
   ],
   "source": [
    "ls ../.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting docking process for gene_id: AGAP000818, ligand: deltamethrin3d\n",
      "receptor file found...\n",
      "Active site coordinates: x=-15.846, y=-23.032, z=-11.293\n",
      "Skipping GNINA docking: log file already exists or ligand file is missing\n",
      "Starting docking process for gene_id: AGAP000818, ligand: deltamethrin3d\n",
      "Receptor file not found. Downloading and preparing AlphaFold PDB for AGAP000818\n",
      "UniProt accession: Q6VFG9\n",
      "UniProt accession: Q6VFH1\n",
      "UniProt accession: Q6VFH2\n",
      "UniProt accession: Q6VFH3\n",
      "UniProt accession: Q8T633\n",
      "selecting longest transcript for gene_id: AGAP000818, Q8T633 : 531\n",
      "┌──────────────────────────────────────────────────────────────────────────────┐\n",
      "│                                 PyRosetta-4                                  │\n",
      "│              Created in JHU by Sergey Lyskov and PyRosetta Team              │\n",
      "│              (C) Copyright Rosetta Commons Member Institutions               │\n",
      "│                                                                              │\n",
      "│ NOTE: USE OF PyRosetta FOR COMMERCIAL PURPOSES REQUIRE PURCHASE OF A LICENSE │\n",
      "│         See LICENSE.PyRosetta.md or email license@uw.edu for details         │\n",
      "└──────────────────────────────────────────────────────────────────────────────┘\n",
      "PyRosetta-4 2024 [Rosetta PyRosetta4.Release.python311.ubuntu 2024.19+release.a34b73c40fe9c61558d566d6a63f803cfb15a4fc 2024-05-02T16:22:03] retrieved from: http://www.pyrosetta.org\n",
      "core.init: Checking for fconfig files in pwd and ./rosetta/flags\n",
      "core.init: Rosetta version: PyRosetta4.Release.python311.ubuntu r381 2024.19+release.a34b73c40f a34b73c40fe9c61558d566d6a63f803cfb15a4fc http://www.pyrosetta.org 2024-05-02T16:22:03\n",
      "core.init: Rosetta extras: []\n",
      "core.init: command: PyRosetta -write_pdb_title_section_records true -write_pdb_link_records true -use_pdb_format_HETNAM_records true -write_pdb_parametric_info true -write_glycan_pdb_codes false -write_seqres_records true -output_pose_energies_table false -database /home/sanj/apps/mambaforge/lib/python3.11/site-packages/pyrosetta/database\n",
      "basic.random.init_random_generator: 'RNG device' seed mode, using '/dev/urandom', seed=925614987 seed_offset=0 real_seed=925614987\n",
      "basic.random.init_random_generator: RandomGenerator:init: Normal mode, seed=925614987 RG_type=mt19937\n",
      "core.chemical.GlobalResidueTypeSet: Finished initializing fa_standard residue type set.  Created 985 residue types\n",
      "core.chemical.GlobalResidueTypeSet: Total time to initialize 0.652274 seconds.\n",
      "core.import_pose.import_pose: File '../../receptors/AGAP000818_raw.pdb' automatically determined to be of type PDB\n",
      "core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015\n",
      "core.scoring.etable: Starting energy table calculation\n",
      "core.scoring.etable: smooth_etable: changing atr/rep split to bottom of energy well\n",
      "core.scoring.etable: smooth_etable: spline smoothing lj etables (maxdis = 6)\n",
      "core.scoring.etable: smooth_etable: spline smoothing solvation etables (max_dis = 6)\n",
      "core.scoring.etable: Finished calculating energy tables.\n",
      "basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/HBPoly1D.csv\n",
      "basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/HBFadeIntervals.csv\n",
      "basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/HBEval.csv\n",
      "basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/DonStrength.csv\n",
      "basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/AccStrength.csv\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/fd/all.ramaProb\n",
      "basic.io.database: Database file opened: scoring/score_functions/rama/fd/prepro.ramaProb\n",
      "basic.io.database: Database file opened: scoring/score_functions/omega/omega_ppdep.all.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/omega/omega_ppdep.gly.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/omega/omega_ppdep.pro.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/omega/omega_ppdep.valile.txt\n",
      "basic.io.database: Database file opened: scoring/score_functions/P_AA_pp/P_AA\n",
      "basic.io.database: Database file opened: scoring/score_functions/P_AA_pp/P_AA_n\n",
      "core.scoring.P_AA: shapovalov_lib::shap_p_aa_pp_smooth_level of 1( aka low_smooth ) got activated.\n",
      "basic.io.database: Database file opened: scoring/score_functions/P_AA_pp/shapovalov/10deg/kappa131/a20.prop\n",
      "core.pack.task: Packer task: initialize from command line()\n",
      "basic.io.database: Database file opened: scoring/score_functions/elec_cp_reps.dat\n",
      "core.scoring.elec.util: Read 40 countpair representative atoms\n",
      "core.pack.dunbrack.RotamerLibrary: shapovalov_lib_fixes_enable option is true.\n",
      "core.pack.dunbrack.RotamerLibrary: shapovalov_lib::shap_dun10_smooth_level of 1( aka lowest_smooth ) got activated.\n",
      "core.pack.dunbrack.RotamerLibrary: Binary rotamer library selected: /home/sanj/apps/mambaforge/lib/python3.11/site-packages/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin\n",
      "core.pack.dunbrack.RotamerLibrary: Using Dunbrack library binary file '/home/sanj/apps/mambaforge/lib/python3.11/site-packages/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin'.\n",
      "core.pack.dunbrack.RotamerLibrary: Dunbrack 2010 library took 0.112276 seconds to load from binary\n",
      "core.pack.pack_rotamers: built 1 rotamers at 1 positions.\n",
      "core.pack.interaction_graph.interaction_graph_factory: Instantiating PDInteractionGraph\n",
      "Mutated residue 224 to I\n",
      "Mutagenesis complete. New structure saved as '../../receptors/AGAP000818_raw_224I.pdb'\n",
      "Successfully downloaded 1TQN.pdb to ./1TQN.pdb\n",
      "INFO:\n",
      "available bundle cache has not been initialized yet\n",
      "INFO:\n",
      "UCSF ChimeraX version: 1.8 (2024-06-10)\n",
      "INFO:\n",
      "© 2016-2024 Regents of the University of California.  All rights reserved.\n",
      "INFO:\n",
      "Executing: open temp_chimerax_script.cxc\n",
      "INFO:\n",
      "Executing: open ../../receptors/AGAP000818_raw_224I.pdb\n",
      "INFO:\n",
      "Chain information for AGAP000818_raw_224I.pdb #1  \n",
      "---  \n",
      "Chain | Description  \n",
      "[A](cxcmd:select /A:1-531 \"Select chain\") | [No description available](cxcmd:sequence chain #1/A \"Show sequence\")  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Executing: open ./1TQN.pdb\n",
      "INFO:\n",
      "_1TQN.pdb_ title:  \n",
      "**Crystal structure of human microsomal P450 3A4** [[more info...]](cxcmd:log\n",
      "metadata #2)  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Chain information for 1TQN.pdb #2  \n",
      "---  \n",
      "Chain | Description | UniProt  \n",
      "[A](cxcmd:select #2/A:28-499 \"Select chain\") | [cytochrome P450 3A4](cxcmd:sequence chain #2/A \"Show sequence\") | [CP3A4_HUMAN](cxcmd:open P08684 from uniprot associate #2/A \"Show annotations\") [22-503](cxcmd:select #2/A:22-503 \"Select sequence\")  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Non-standard residues in 1TQN.pdb #2  \n",
      "---  \n",
      "[HEM](cxcmd:sel :HEM \"select residue\") — [protoporphyrin IX containing\n",
      "Fe](http://www.rcsb.org/ligand/HEM \"show residue info\") (HEME)  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Executing: matchmaker #1 to #2\n",
      "INFO:\n",
      "Parameters  \n",
      "---  \n",
      "Chain pairing | bb  \n",
      "Alignment algorithm | Needleman-Wunsch  \n",
      "Similarity matrix | BLOSUM-62  \n",
      "SS fraction | 0.3  \n",
      "Gap open (HH/SS/other) | 18/18/6  \n",
      "Gap extend | 1  \n",
      "SS matrix |  |  | H | S | O  \n",
      "---|---|---|---  \n",
      "H | 6 | -9 | -6  \n",
      "S |  | 6 | -6  \n",
      "O |  |  | 4  \n",
      "Iteration cutoff | 2  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Matchmaker 1TQN.pdb, chain A (#2) with AGAP000818_raw_224I.pdb, chain A (#1), sequence alignment score = 917.1\n",
      "STATUS:\n",
      "Matchmaker 1TQN.pdb, chain A (#2) with AGAP000818_raw_224I.pdb, chain A (#1), sequence alignment score = 917.1\n",
      "INFO:\n",
      "RMSD between 314 pruned atom pairs is 1.012 angstroms; (across all 456 pairs: 3.044)\n",
      "STATUS:\n",
      "RMSD between 314 pruned atom pairs is 1.012 angstroms; (across all 456 pairs: 3.044)\n",
      "INFO:\n",
      "\n",
      "INFO:\n",
      "Executing: close #2\n",
      "INFO:\n",
      "Executing: save ../../receptors/AGAP000818_raw_224I.pdb models #1\n",
      "INFO:\n",
      "Executing: exit\n",
      "STATUS:\n",
      "Exiting ...\n",
      "ChimeraX alignment completed. Result saved to ../../receptors/AGAP000818_raw_224I.pdb\n",
      "HEM coordinates inserted successfully. Output written to ../../receptors/AGAP000818_raw_224I.pdb\n",
      "Number of HEM lines inserted: 43\n",
      "Downloaded and protonated AlphaFold PDB for AGAP000818 (UniProt: Q8T633) to ../../receptors/AGAP000818_224I.pdbqt\n",
      "Active site coordinates: x=-15.846, y=-23.032, z=-11.293\n",
      "Preparing to run GNINA for docking...\n",
      "              _             \n",
      "             (_)            \n",
      "   __ _ _ __  _ _ __   __ _ \n",
      "  / _` | '_ \\| | '_ \\ / _` |\n",
      " | (_| | | | | | | | | (_| |\n",
      "  \\__, |_| |_|_|_| |_|\\__,_|\n",
      "   __/ |                    \n",
      "  |___/                     \n",
      "\n",
      "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n",
      "gnina is based on smina and AutoDock Vina.\n",
      "Please cite appropriately.\n",
      "\n",
      "WARNING: No GPU detected. CNN scoring will be slow.\n",
      "Recommend running with single model (--cnn crossdock_default2018)\n",
      "or without cnn scoring (--cnn_scoring=none).\n",
      "\n",
      "Commandline: ./gnina -r ../../receptors/AGAP000818_224I.pdbqt -l ../../ligands/deltamethrin3d.pdbqt --center_x -15.846 --center_y -23.032 --center_z -11.293 --size_x 20 --size_y 20 --size_z 20 -o ../../docked/AGAP000818_224I_deltamethrin3d.sdf --log ../../logs/AGAP000818_224I_deltamethrin3d.log --seed 0\n",
      "Using random seed: 0\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      " | pose 0 | initial pose not within box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | initial pose not within box\n",
      "\n",
      "mode |  affinity  |  intramol  |    CNN     |   CNN\n",
      "     | (kcal/mol) | (kcal/mol) | pose score | affinity\n",
      "-----+------------+------------+------------+----------\n",
      "    1       -5.82        1.78       0.4462      6.882\n",
      "    2       -4.40        3.67       0.4107      6.731\n",
      "    3       -7.40        1.48       0.4087      6.959\n",
      "    4       -4.79        0.62       0.4079      6.996\n",
      "    5       -6.56        0.18       0.3905      6.932\n",
      "    6       -6.00        0.07       0.3826      6.914\n",
      "    7       -5.68        1.34       0.3767      6.879\n",
      "    8       -6.19       -0.26       0.3357      6.917\n",
      "    9       -7.53        1.69       0.3332      6.762\n"
     ]
    }
   ],
   "source": [
    "import imove\n",
    "for mutagenesis_dict in [None, {224:'I'}]:\n",
    "    docked = imove.dock(\"AGAP000818\",\n",
    "                        \"deltamethrin3d\", \n",
    "                        wkdir=\"../../\", \n",
    "                        smarts_for_distance=None,\n",
    "                        verbose=True, \n",
    "                        mutagenesis_dict=mutagenesis_dict,\n",
    "                        override_motif=None,\n",
    "                        p450=True\n",
    "                       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting docking process for gene_id: AGAP002862, ligand: deltamethrin3d\n",
      "Receptor file not found. Downloading and preparing AlphaFold PDB for AGAP002862\n",
      "Successfully downloaded 1TQN.pdb to ./1TQN.pdb\n",
      "INFO:\n",
      "available bundle cache has not been initialized yet\n",
      "INFO:\n",
      "UCSF ChimeraX version: 1.8 (2024-06-10)\n",
      "INFO:\n",
      "© 2016-2024 Regents of the University of California.  All rights reserved.\n",
      "INFO:\n",
      "Executing: open temp_chimerax_script.cxc\n",
      "INFO:\n",
      "Executing: open ../../receptors/AGAP002862_raw.pdb\n",
      "INFO:\n",
      "_AGAP002862_raw.pdb_ title:  \n",
      "**Alphafold monomer V2.0 prediction for AGAP002862-Pa (Q8MZK0)**\n",
      "[[more info...]](cxcmd:log metadata #1)  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Chain information for AGAP002862_raw.pdb #1  \n",
      "---  \n",
      "Chain | Description | UniProt  \n",
      "[A](cxcmd:select /A:1-505 \"Select chain\") | [AGAP002862-Pa](cxcmd:sequence chain #1/A \"Show sequence\") | [Q8MZK0_ANOGA](cxcmd:open Q8MZK0 from uniprot associate #1/A \"Show annotations\") [1-505](cxcmd:select #1/A:1-505 \"Select sequence\")  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Executing: open ./1TQN.pdb\n",
      "INFO:\n",
      "_1TQN.pdb_ title:  \n",
      "**Crystal structure of human microsomal P450 3A4** [[more info...]](cxcmd:log\n",
      "metadata #2)  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Chain information for 1TQN.pdb #2  \n",
      "---  \n",
      "Chain | Description | UniProt  \n",
      "[A](cxcmd:select #2/A:28-499 \"Select chain\") | [cytochrome P450 3A4](cxcmd:sequence chain #2/A \"Show sequence\") | [CP3A4_HUMAN](cxcmd:open P08684 from uniprot associate #2/A \"Show annotations\") [22-503](cxcmd:select #2/A:22-503 \"Select sequence\")  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Non-standard residues in 1TQN.pdb #2  \n",
      "---  \n",
      "[HEM](cxcmd:sel :HEM \"select residue\") — [protoporphyrin IX containing\n",
      "Fe](http://www.rcsb.org/ligand/HEM \"show residue info\") (HEME)  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Executing: matchmaker #1 to #2\n",
      "INFO:\n",
      "Parameters  \n",
      "---  \n",
      "Chain pairing | bb  \n",
      "Alignment algorithm | Needleman-Wunsch  \n",
      "Similarity matrix | BLOSUM-62  \n",
      "SS fraction | 0.3  \n",
      "Gap open (HH/SS/other) | 18/18/6  \n",
      "Gap extend | 1  \n",
      "SS matrix |  |  | H | S | O  \n",
      "---|---|---|---  \n",
      "H | 6 | -9 | -6  \n",
      "S |  | 6 | -6  \n",
      "O |  |  | 4  \n",
      "Iteration cutoff | 2  \n",
      "  \n",
      "\n",
      "INFO:\n",
      "Matchmaker 1TQN.pdb, chain A (#2) with AGAP002862_raw.pdb, chain A (#1), sequence alignment score = 966.7\n",
      "STATUS:\n",
      "Matchmaker 1TQN.pdb, chain A (#2) with AGAP002862_raw.pdb, chain A (#1), sequence alignment score = 966.7\n",
      "INFO:\n",
      "RMSD between 278 pruned atom pairs is 0.970 angstroms; (across all 457 pairs: 4.155)\n",
      "STATUS:\n",
      "RMSD between 278 pruned atom pairs is 0.970 angstroms; (across all 457 pairs: 4.155)\n",
      "INFO:\n",
      "\n",
      "INFO:\n",
      "Executing: close #2\n",
      "INFO:\n",
      "Executing: save ../../receptors/AGAP002862_raw.pdb models #1\n",
      "INFO:\n",
      "Executing: exit\n",
      "STATUS:\n",
      "Exiting ...\n",
      "ChimeraX alignment completed. Result saved to ../../receptors/AGAP002862_raw.pdb\n",
      "HEM coordinates inserted successfully. Output written to ../../receptors/AGAP002862_raw.pdb\n",
      "Number of HEM lines inserted: 43\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "==============================\n",
      "*** Open Babel Warning  in PerceiveBondOrders\n",
      "  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders (title is ../../receptors/AGAP002862_raw.pdb)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloaded and protonated AlphaFold PDB for AGAP002862 (UniProt: Q8MZK0) to ../../receptors/AGAP002862.pdbqt\n",
      "Active site coordinates: x=-15.846, y=-23.032, z=-11.293\n",
      "Preparing to run GNINA for docking...\n",
      "              _             \n",
      "             (_)            \n",
      "   __ _ _ __  _ _ __   __ _ \n",
      "  / _` | '_ \\| | '_ \\ / _` |\n",
      " | (_| | | | | | | | | (_| |\n",
      "  \\__, |_| |_|_|_| |_|\\__,_|\n",
      "   __/ |                    \n",
      "  |___/                     \n",
      "\n",
      "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n",
      "gnina is based on smina and AutoDock Vina.\n",
      "Please cite appropriately.\n",
      "\n",
      "WARNING: No GPU detected. CNN scoring will be slow.\n",
      "Recommend running with single model (--cnn crossdock_default2018)\n",
      "or without cnn scoring (--cnn_scoring=none).\n",
      "\n",
      "Commandline: ./gnina -r ../../receptors/AGAP002862.pdbqt -l ../../ligands/deltamethrin3d.pdbqt --center_x -15.846 --center_y -23.032 --center_z -11.293 --size_x 20 --size_y 20 --size_z 20 -o ../../docked/AGAP002862_deltamethrin3d.sdf --log ../../logs/AGAP002862_deltamethrin3d.log --seed 0\n",
      "Using random seed: 0\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "|----|----|----|----|----|----|----|----|----|----|\n",
      "***************************************************\n",
      " | pose 0 | initial pose not within box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | initial pose not within box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | initial pose not within box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      " | pose 0 | ligand outside box\n",
      "\n",
      "mode |  affinity  |  intramol  |    CNN     |   CNN\n",
      "     | (kcal/mol) | (kcal/mol) | pose score | affinity\n",
      "-----+------------+------------+------------+----------\n",
      "    1       -8.02        1.80       0.7596      7.364\n",
      "    2       -7.36       -0.89       0.7439      7.335\n",
      "    3       -7.91        1.37       0.7012      7.244\n",
      "    4       -7.83        0.17       0.6836      7.064\n",
      "    5       -8.22       -0.78       0.6472      7.173\n",
      "    6       -6.78        0.80       0.6105      7.173\n",
      "    7       -7.64        0.25       0.5891      7.019\n",
      "    8       -8.85        0.24       0.5867      7.104\n",
      "    9       -7.54        0.87       0.5860      7.096\n"
     ]
    }
   ],
   "source": [
    "docked = imove.dock(\"AGAP002862\",\n",
    "                    \"deltamethrin3d\", \n",
    "                    wkdir=\"../../\", \n",
    "                    smarts_for_distance=None,\n",
    "                    verbose=True, \n",
    "                    mutagenesis_dict=None,#{155:'N'},\n",
    "                    override_motif=None,\n",
    "                    p450=True\n",
    "                   )"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "include_colab_link": true,
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
