{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "authorship_tag": "ABX9TyP+RWEX5LBtgpvU7SGuA4te",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/sanjaynagi/AnoFold/blob/main/notebooks/AnoFold.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%pip install py3Dmol\n",
        "\n",
        "!wget https://github.com/gnina/gnina/releases/download/v1.1/gnina\n",
        "!chmod +x gnina\n",
        "!./gnina --version"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "x5wID6CTbzRB",
        "outputId": "5f4c81b4-b287-40de-b725-cd860b73901d"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting py3Dmol\n",
            "  Downloading py3Dmol-2.3.0-py2.py3-none-any.whl.metadata (1.9 kB)\n",
            "Downloading py3Dmol-2.3.0-py2.py3-none-any.whl (7.0 kB)\n",
            "Installing collected packages: py3Dmol\n",
            "Successfully installed py3Dmol-2.3.0\n",
            "--2024-08-19 18:26:37--  https://github.com/gnina/gnina/releases/download/v1.1/gnina\n",
            "Resolving github.com (github.com)... 140.82.116.3\n",
            "Connecting to github.com (github.com)|140.82.116.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/45548146/bc227ff8-7934-457d-95b3-eab58982638a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240819%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240819T182637Z&X-Amz-Expires=300&X-Amz-Signature=8c6227f5f32b6657488105d385b384e3d6e90bc338fe269f455e92833b71654e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=45548146&response-content-disposition=attachment%3B%20filename%3Dgnina&response-content-type=application%2Foctet-stream [following]\n",
            "--2024-08-19 18:26:37--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/45548146/bc227ff8-7934-457d-95b3-eab58982638a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240819%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240819T182637Z&X-Amz-Expires=300&X-Amz-Signature=8c6227f5f32b6657488105d385b384e3d6e90bc338fe269f455e92833b71654e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=45548146&response-content-disposition=attachment%3B%20filename%3Dgnina&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 306470832 (292M) [application/octet-stream]\n",
            "Saving to: ‘gnina’\n",
            "\n",
            "gnina               100%[===================>] 292.27M   102MB/s    in 2.9s    \n",
            "\n",
            "2024-08-19 18:26:41 (102 MB/s) - ‘gnina’ saved [306470832/306470832]\n",
            "\n",
            "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "id": "2kA4VyCKTm5-"
      },
      "outputs": [],
      "source": [
        "import py3Dmol\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import plotly.express as px\n",
        "import requests\n",
        "import os\n",
        "import re\n",
        "import glob\n",
        "\n",
        "def load_gene_metadata(gene_id):\n",
        "\n",
        "  df_genes = pd.read_csv(\"https://raw.githubusercontent.com/sanjaynagi/AnoExpress/main/resources/AgamP4.annots.tsv\", sep=\"\\t\")\n",
        "  df_genes = df_genes.query(\"GeneID == @gene_id\")\n",
        "  return df_genes.iloc[0].values\n",
        "\n",
        "\n",
        "def vectorbase_to_uniprot(gene_id):\n",
        "    url = \"https://rest.uniprot.org/idmapping/run\"\n",
        "    data = {\n",
        "        \"from\": \"VEuPathDB\",\n",
        "        \"to\": \"UniProtKB\",\n",
        "        \"ids\": f\"VectorBase:{gene_id}\"\n",
        "    }\n",
        "\n",
        "    response = requests.post(url, data=data)\n",
        "    response.raise_for_status()\n",
        "    job_id = response.json()[\"jobId\"]\n",
        "\n",
        "    status_url = f\"https://rest.uniprot.org/idmapping/status/{job_id}\"\n",
        "    while True:\n",
        "        status_response = requests.get(status_url)\n",
        "        status_response.raise_for_status()\n",
        "        status = status_response.json()\n",
        "        if \"jobStatus\" in status and status[\"jobStatus\"] in (\"RUNNING\", \"NEW\"):\n",
        "            continue\n",
        "        elif \"results\" in status or \"failedIds\" in status:\n",
        "            break\n",
        "\n",
        "    results_url = f\"https://rest.uniprot.org/idmapping/stream/{job_id}\"\n",
        "    results_response = requests.get(results_url)\n",
        "    results_response.raise_for_status()\n",
        "    results = results_response.json()\n",
        "\n",
        "    if \"results\" in results and results[\"results\"]:\n",
        "        return results[\"results\"][0][\"to\"]\n",
        "    else:\n",
        "        return None\n",
        "\n",
        "def download_alphafold_pdb(gene_id, output_dir='.'):\n",
        "    # Convert VectorBase GeneID to UniProt accession\n",
        "    uniprot_accession = vectorbase_to_uniprot(gene_id)\n",
        "    if uniprot_accession is None:\n",
        "        print(f\"No UniProt accession found for GeneID: {gene_id}\")\n",
        "        return None\n",
        "\n",
        "    # Download the PDB file\n",
        "    response = requests.get(f\"https://alphafold.ebi.ac.uk/files/AF-{uniprot_accession}-F1-model_v4.pdb\")\n",
        "    if response.status_code == 200:\n",
        "        # Create the output directory if it doesn't exist\n",
        "        os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "        # Save the PDB file\n",
        "        output_file = os.path.join(output_dir, f\"{gene_id}.pdb\")\n",
        "        with open(output_file, 'wb') as f:\n",
        "            f.write(response.content)\n",
        "\n",
        "        print(f\"Downloaded AlphaFold PDB for {gene_id} (UniProt: {uniprot_accession}) to {output_file}\")\n",
        "        return output_file\n",
        "    else:\n",
        "        print(f\"Failed to download AlphaFold PDB for {gene_id} (UniProt: {uniprot_accession}). Status code: {response.status_code}\")\n",
        "        return None\n",
        "\n",
        "amino_acid_map = {\n",
        "    'ALA': 'A', 'ARG': 'R', 'ASN': 'N', 'ASP': 'D', 'CYS': 'C',\n",
        "    'GLN': 'Q', 'GLU': 'E', 'GLY': 'G', 'HIS': 'H', 'ILE': 'I',\n",
        "    'LEU': 'L', 'LYS': 'K', 'MET': 'M', 'PHE': 'F', 'PRO': 'P',\n",
        "    'SER': 'S', 'THR': 'T', 'TRP': 'W', 'TYR': 'Y', 'VAL': 'V'\n",
        "    }\n",
        "\n",
        "# motif_map = {'P450':,'blah',\n",
        "#              'COE':['[LIV].G.S.G', 4, \"0\"]}\n",
        "\n",
        "def load_pdb(pdb_path):\n",
        "    # Read and process the PDB file\n",
        "    df_pdb = pd.read_csv(pdb_path, skiprows=85, header=None)[0].str.split(\"\\s+\", expand=True).iloc[:-3,:-1]\n",
        "    df_pdb.columns = [\n",
        "        \"record\", \"atom\", \"atom_name\", \"amino_acid\", \"chain_identifier\", \"codon\",\n",
        "        \"x\", \"y\", \"z\", \"occupancy\", \"b-factor\", \"element_symbol\",\n",
        "    ]\n",
        "    df_pdb = df_pdb.query(\"record == 'ATOM'\")\n",
        "    return df_pdb\n",
        "\n",
        "def pdb_to_active_site_coords(pdb_path, target_motif, target_codon_in_motif, target_molecule, flanking_size=5):\n",
        "    df_pdb = load_pdb(pdb_path)\n",
        "\n",
        "    # Extract unique amino acids\n",
        "    aa3 = df_pdb[['amino_acid', 'codon']].drop_duplicates()['amino_acid']\n",
        "\n",
        "    convert = np.vectorize(lambda x: amino_acid_map.get(x, x))\n",
        "    aa1 = convert(aa3)\n",
        "\n",
        "    # Find all occurrences of the motif\n",
        "    aa_sequence = ''.join(aa1)\n",
        "    motif_matches = list(re.finditer(target_motif, aa_sequence))\n",
        "\n",
        "    results = []\n",
        "    for match in motif_matches:\n",
        "        start, end = match.start(), match.end()\n",
        "\n",
        "        # Get flanking region\n",
        "        flanking_start = max(0, start - flanking_size)\n",
        "        flanking_end = min(len(aa_sequence), end + flanking_size)\n",
        "        flanking = aa_sequence[flanking_start:start] + '[' + aa_sequence[start:end] + ']' + aa_sequence[end:flanking_end]\n",
        "\n",
        "        print(f\"Motif detected in PDB at codon {start+1}:{end} = {aa_sequence[start:end]}\")\n",
        "        print(f\"Flanking sequence: {flanking}\")\n",
        "\n",
        "        # Find coordinates for the target molecule\n",
        "        target_idx = str(start + 1 + target_codon_in_motif)\n",
        "        coords = df_pdb.query(\"codon == @target_idx and atom_name == @target_molecule\")\n",
        "\n",
        "        if not coords.empty:\n",
        "            coord_array = coords.drop_duplicates('atom_name')[['x', 'y', 'z']].to_numpy()[0].astype(float)\n",
        "            results.append({\n",
        "                'start': start + 1,\n",
        "                'end': end,\n",
        "                'motif': aa_sequence[start:end],\n",
        "                'flanking': flanking,\n",
        "                'coordinates': coord_array,\n",
        "                'molecule_number': int(coords['atom'].values[0])\n",
        "            })\n",
        "\n",
        "    if len(results) > 1:\n",
        "        print(f\"Warning, multiple {len(results)} matching motifs found\")\n",
        "\n",
        "    return results\n",
        "\n",
        "import requests\n",
        "def download_ligand(ligand_name, repo_url=\"https://raw.githubusercontent.com/sanjaynagi/AnoFold/main/ligands\", save_path=\"ligands\"):\n",
        "    \"\"\"\n",
        "    Download a ligand file from the GitHub repository.\n",
        "\n",
        "    Parameters:\n",
        "    - ligand_name (str): The name of the ligand (without the .pdbqt extension).\n",
        "    - repo_url (str): The base URL of the GitHub repository's raw content.\n",
        "    - save_path (str): The local directory where the file will be saved. Default is current directory.\n",
        "\n",
        "    Example:\n",
        "    download_ligand(\"deltamethrin\", \"https://raw.githubusercontent.com/sanjaynagi/AnoFold/main/ligands/\")\n",
        "    \"\"\"\n",
        "    os.makedirs(save_path, exist_ok=True)\n",
        "    file_url = f\"{repo_url}/{ligand_name}.pdbqt\"\n",
        "    file_path = f\"{save_path}/{ligand_name}.pdbqt\"\n",
        "\n",
        "    try:\n",
        "        response = requests.get(file_url)\n",
        "        response.raise_for_status()  # Check if the request was successful\n",
        "\n",
        "        with open(file_path, 'wb') as file:\n",
        "            file.write(response.content)\n",
        "        print(f\"Ligand '{ligand_name}.pdbqt' downloaded successfully.\\n\")\n",
        "    except requests.exceptions.RequestException as e:\n",
        "        print(f\"Failed to download ligand: {e}\")\n",
        "\n",
        "def view_pymol(receptor, ligand, docked, receptor_highlight=None, sticks=False):\n",
        "  v = py3Dmol.view()\n",
        "  v.addModel(open(receptor).read())\n",
        "  if sticks:\n",
        "    v.setStyle({'cartoon':{},'stick':{'radius':.1}})\n",
        "  else:\n",
        "    v.setStyle({'cartoon':{}})\n",
        "  if receptor_highlight:\n",
        "    for i in range(receptor_highlight-3, receptor_highlight+3):\n",
        "      v.setStyle({'model': -1, 'serial': i}, {\"cartoon\": {'color': 'yellow'}, 'stick':{'radius':.3, 'color':'yellow'}})\n",
        "  v.addModel(open(ligand).read())\n",
        "  v.setStyle({'model':1},{'stick':{'colorscheme':'dimgrayCarbon','radius':.125}})\n",
        "  v.addModelsAsFrames(open(docked).read())\n",
        "  v.setStyle({'model':2},{'stick':{'colorscheme':'greenCarbon'}})\n",
        "  v.zoomTo({'model':1})\n",
        "  v.rotate(90)\n",
        "  v.animate({'interval':5000})\n",
        "  return v\n",
        "\n",
        "def plot_molecular_docking(gene_id, ligand, custom_active_site_motif=None, custom_target_codon_in_motif=None, custom_target_molecule=None, **kwargs):\n",
        "\n",
        "  pdb_path  = f\"receptors/{gene_id}.pdb\"\n",
        "  ligand_path = f\"ligands/{ligand}.pdbqt\"\n",
        "\n",
        "  # get_gene_info\n",
        "  if not custom_active_site_motif:\n",
        "    assert gene_id.startswith(\"AGAP\"), \"custom active site motif necessary for other organisms\"\n",
        "    if gene_id.startswith(\"AGAP\"):\n",
        "      gene_id, gene_name, gene_desc = load_gene_metadata(gene_id)\n",
        "      print(f\"Running molecular docking for {gene_id} | {gene_name} ({gene_desc}),  and {ligand}.\")\n",
        "      if 'P450' in gene_desc:\n",
        "        custom_active_site_motif, custom_target_molecule, custom_target_codon_in_motif = ('blah', 'C', 3)\n",
        "      elif 'esterase' in gene_desc:\n",
        "        custom_active_site_motif, custom_target_molecule, custom_target_codon_in_motif = ('[LIV].G.S.G', 'O', 4)\n",
        "      else:\n",
        "        assert \"Unknown gene family, custom motif required\"\n",
        "\n",
        "  # download inputs\n",
        "  if not os.path.exists(f\"receptors/{gene_id}.pdb\"):\n",
        "    download_alphafold_pdb(gene_id, output_dir=\"receptors\")\n",
        "  if not os.path.exists(f\"ligands/{ligand}.pdbqt\"):\n",
        "    download_ligand(ligand)\n",
        "\n",
        "  # get active site coords\n",
        "  res = pdb_to_active_site_coords(\n",
        "      pdb_path=pdb_path,\n",
        "      target_motif=custom_active_site_motif, target_molecule=custom_target_molecule, target_codon_in_motif=custom_target_codon_in_motif)\n",
        "  x,y,z = res[0]['coordinates']\n",
        "\n",
        "  # run gnina\n",
        "  if not os.path.exists(f\"docking/{gene_id}_{ligand}.log\"):\n",
        "    os.makedirs(\"docking\", exist_ok=True)\n",
        "    !./gnina -r {pdb_path} -l {ligand_path} --center_x {x} --center_y {y} --center_z {z} --size_x 20 --size_y 20 --size_z 20 -o docking/{gene_id}_{ligand}.sdf --log docking/{gene_id}_{ligand}.log --seed 0\n",
        "\n",
        "\n",
        "  # plot\n",
        "  return view_pymol(receptor=pdb_path, ligand=ligand_path, docked=f\"docking/{gene_id}_{ligand}.sdf\", receptor_highlight=res[0]['molecule_number'], **kwargs)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## AnoFold - molecular docking with alphafold and gnina"
      ],
      "metadata": {
        "id": "FmfNINymb3_o"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "load_gene_metadata('AGAP006227').iloc[0].values"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "arJf-Brt5T5w",
        "outputId": "3992fbc5-4072-46d2-e2ae-1a0e1a414d67"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['AGAP006227', 'COEAE1F',\n",
              "       'alpha esterase [Source:VB Community Annotation]'], dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "plot_molecular_docking(\"AGAP006227\", ligand='malathion'),# active_site_motif='[LIV].G.S.G', target_codon_in_motif=4, target_molecule=\"O\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EngoyYsExeUr",
        "outputId": "e325f7d8-2760-4063-a58e-90770c6abddd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Running molecular docking for AGAP006227 | COEAE1F (a alpha esterase [Source:VB Community Annotation]),  and malathion.\n",
            "Downloaded AlphaFold PDB for AGAP006227 (UniProt: Q7PPA9) to receptors/AGAP006227.pdb\n",
            "Ligand 'malathion.pdbqt' downloaded successfully.\n",
            "\n",
            "Motif detected in PDB at codon 188:194 = LFGESAG\n",
            "Flanking sequence: PDNVT[LFGESAG]GCSVH\n",
            "              _             \n",
            "             (_)            \n",
            "   __ _ _ __  _ _ __   __ _ \n",
            "  / _` | '_ \\| | '_ \\ / _` |\n",
            " | (_| | | | | | | | | (_| |\n",
            "  \\__, |_| |_|_|_| |_|\\__,_|\n",
            "   __/ |                    \n",
            "  |___/                     \n",
            "\n",
            "gnina v1.1 master:e4cb380+   Built Dec 18 2023.\n",
            "gnina is based on smina and AutoDock Vina.\n",
            "Please cite appropriately.\n",
            "\n",
            "WARNING: No GPU detected. CNN scoring will be slow.\n",
            "Recommend running with single model (--cnn crossdock_default2018)\n",
            "or without cnn scoring (--cnn_scoring=none).\n",
            "\n",
            "Commandline: ./gnina -r receptors/AGAP006227.pdb -l ligands/malathion.pdbqt --center_x 1.481 --center_y 1.619 --center_z -1.237 --size_x 20 --size_y 20 --size_z 20 -o docking/AGAP006227_malathion.sdf --log docking/AGAP006227_malathion.log --seed 0\n",
            "==============================\n",
            "*** Open Babel Warning  in Init\n",
            "  Cannot initialize database 'space-groups.txt' which may cause further errors.\n",
            "==============================\n",
            "*** Open Babel Warning  in PerceiveBondOrders\n",
            "  Failed to kekulize aromatic bonds in OBMol::PerceiveBondOrders\n",
            "\n",
            "Using random seed: 0\n",
            "\n",
            "0%   10   20   30   40   50   60   70   80   90   100%\n",
            "|----|----|----|----|----|----|----|----|----|----|\n",
            "**************************"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "gene_id = \"AGAP006227\"\n",
        "ligand = \"malathion\"\n",
        "\n",
        "pdb_path = f\"receptors/{gene_id}.pdb\"\n",
        "ligand_path = f\"ligands/{ligand}.pdbqt\""
      ],
      "metadata": {
        "id": "VSmkYZLi8pc6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Iesomu5OHATQ"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}